torch_geometric/explain/algorithm/pgm_explainer.py

```diff
@@ -0,0 +1,440 @@
+            feature_matrix: torch.Tensor,
+            node_idx: int,
+            pertubation_mode="randint",
+            is_random: bool = False,
+            is_pertubation_scaled: bool = False,
+            perturb_feature_list=None  # indexes of features being pertubed
+    ):
+        r"""
+        pertub node feature matrix. This allows for checking how much influence
+        neighbouring nodes has on the output
+        Args:
+            feature_matrix: node feature matrix of the input graph
+                of shape [num_nodes, num_features]
+            node_idx: index of the node we are calculating the explanation for
+            pertubation_mode: how to pertube the features. must be one of
+                [random, zero, mean, max]
+            is_pertubation_scaled: whether to scale the pertubed matrix
+                with the original feature matrix
+
+        Returns:
+
+        """
+        # return a random perturbed feature matrix
+        # random = is_random for nothing, True for random.randint.
+        # is_pertubation_scaled=True, "scale" for scaling with original feature
+
+        X_perturb = feature_matrix.detach().clone()
+        perturb_array = X_perturb[node_idx].clone()
+        epsilon = 0.05 * torch.max(feature_matrix, dim=0).values
+
+        if is_random:
+            if not is_pertubation_scaled:
+                if pertubation_mode == "randint":
+                    perturb_array = torch.randint(
+                        high=2, size=X_perturb[node_idx].shape[0])
+                # graph explainers
+                elif pertubation_mode in ("mean", "zero", "max", "uniform"):
+                    for i in range(perturb_array.shape[0]):
+                        if i in perturb_feature_list:
+                            if pertubation_mode == "mean":
+                                perturb_array[i] = torch.mean(
+                                    feature_matrix[:, i])
+                            elif pertubation_mode == "zero":
+                                perturb_array[i] = 0
+                            elif pertubation_mode == "max":
+                                perturb_array[i] = torch.max(feature_matrix[:,
+                                                                            i])
+                            elif pertubation_mode == "uniform":
+                                perturb_array[
+                                    i] = perturb_array[i] + torch.tensor(
+                                        np.random.uniform(
+                                            low=-epsilon[i], high=epsilon[i]))
+                                if perturb_array[i] < 0:
+                                    perturb_array[i] = 0
+                                elif perturb_array[i] > torch.max(
+                                        feature_matrix, dim=0)[i]:
+                                    perturb_array[i] = torch.max(
+                                        feature_matrix, dim=0)[i]
```

Comment:
Probably we don’t need the for loop here. Can we use `perturb_feature_list` to access the values in `perturb_array` that need change by` perturb_array[perturb_feature_list]`? Then we can apply different operations according to `pertubation_mode` directly over the tensor (`perturb_array[perturb_feature_list]`).